File: Python3/416. Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 != 0:
            return False
        
        subtotal = total // 2
        dp = [False] * (subtotal + 1)
        dp[0] = True
        
        for n in nums:
            for j in range(subtotal, n - 1, -1):
                dp[j] = dp[j] or dp[j - n]
                
        return dp[subtotal]
        

File: Java/416. Partition Equal Subset Sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int total = 0;
        for (int n: nums) {
            total += n;
        }
        
        if (total % 2 != 0) {
            return false;
        }
        
        int subtotal = total / 2;
        boolean dp[] = new boolean[subtotal + 1];
        dp[0] = true;
        
        for (int n: nums) {
            for (int j=subtotal; j>=n; j--) {
                dp[j] |= dp[j - n];
            }
        }
        
        return dp[subtotal];
    }
}